presto-main/src/main/java/com/facebook/presto/execution/TaskExecutor.java

```diff
@@ -134,28 +141,41 @@
     private final TimeStat blockedQuantaWallTime = new TimeStat(MICROSECONDS);
     private final TimeStat unblockedQuantaWallTime = new TimeStat(MICROSECONDS);
 
+    private final TaskExectorController taskExectorController;
+
+    private final Duration runnerThreadsAdjustmentInterval;
+    @GuardedBy("this")
+    private int targetRunnerThreads;
+    @GuardedBy("this")
+    private int runnerThreads;
+
     private volatile boolean closed;
 
     @Inject
     public TaskExecutor(TaskManagerConfig config)
     {
-        this(requireNonNull(config, "config is null").getMaxWorkerThreads(), config.getMinDrivers());
+        this(createTaskExecutorController(requireNonNull(config, "config is null")), config.getWorkerThreadsAdjustmentInterval(), config.getMinDrivers());
     }
 
-    public TaskExecutor(int runnerThreads, int minDrivers)
+    public TaskExecutor(TaskExectorController controller, Duration workerThreadsAdjustmentInterval, int minDrivers)
     {
-        this(runnerThreads, minDrivers, Ticker.systemTicker());
+        this(controller, workerThreadsAdjustmentInterval, minDrivers, Ticker.systemTicker());
     }
 
     @VisibleForTesting
-    public TaskExecutor(int runnerThreads, int minDrivers, Ticker ticker)
+    public TaskExecutor(TaskExectorController controller, Duration workerThreadsAdjustmentInterval, int minDrivers, Ticker ticker)
     {
-        checkArgument(runnerThreads > 0, "runnerThreads must be at least 1");
+        requireNonNull(controller, "controll is null");
+        requireNonNull(workerThreadsAdjustmentInterval, "workerThreadsAdjustmentInterval is null");
 
         // we manages thread pool size directly, so create an unlimited pool
         this.executor = newCachedThreadPool(threadsNamed("task-processor-%s"));
         this.executorMBean = new ThreadPoolExecutorMBean((ThreadPoolExecutor) executor);
-        this.runnerThreads = runnerThreads;
+
+        this.taskExectorController = controller;
+        this.targetRunnerThreads = 0;
+        this.runnerThreads = 0;
```

Comment:
I do not like the tracking of number of currently running threads on separate variable.
It would nicer to check how many Tasks are actually running in TaskExecutor instead, to have just one source of truth.
Not sure if that is possible.